File: packages/typescript/src/http.ts
import { TossInvestApiError, TossInvestConnectionError } from './errors.js';
import type { TossInvestResponseMeta } from './types.js';

export type FetchLike = typeof fetch;

export interface HttpRequest {
  method: string;
  url: string;
  headers: Headers;
  body?: BodyInit;
  timeoutMs: number;
}

export interface HttpSuccess<TData> {
  data: TData;
  response: TossInvestResponseMeta;
}

export function buildUrl(
  baseUrl: string,
  path: string,
  query: Record<string, string | number | boolean | null | undefined> = {},
): string {
  const url = new URL(path, baseUrl.endsWith('/') ? baseUrl : `${baseUrl}/`);

  for (const [key, value] of Object.entries(query)) {
    if (value !== undefined && value !== null) {
      url.searchParams.set(key, String(value));
    }
  }

  return url.toString();
}

export function encodeForm(data: Record<string, string>): URLSearchParams {
  const form = new URLSearchParams();

  for (const [key, value] of Object.entries(data)) {
    form.set(key, value);
  }

  return form;
}

export function createTimeoutSignal(timeoutMs: number): AbortSignal {
  return AbortSignal.timeout(timeoutMs);
}

export async function requestJson<TData>(
  fetchImpl: FetchLike,
  request: HttpRequest,
): Promise<HttpSuccess<TData>> {
  let response: Response;

  try {
    const init: RequestInit = {
      method: request.method,
      headers: request.headers,
      signal: createTimeoutSignal(request.timeoutMs),
    };

    if (request.body !== undefined) {
      init.body = request.body;
    }

    response = await fetchImpl(request.url, init);
  } catch (error) {
    throw new TossInvestConnectionError(
      'Toss Invest API request failed before receiving a response',
      error,
    );
  }

  const body = await parseJsonBody(response);
  const meta = {
    status: response.status,
    headers: response.headers,
    requestId: response.headers.get('x-request-id') ?? undefined,
  };

  if (!response.ok) {
    const requestId = meta.requestId ?? extractBodyRequestId(body);

    throw new TossInvestApiError({
      status: response.status,
      headers: response.headers,
      body,
      code: extractErrorCode(body),
      message: extractErrorMessage(body),
      requestId,
    });
  }

  return {
    data: body as TData,
    response: meta,
  };
}

async function parseJsonBody(response: Response): Promise<unknown> {
  const text = await response.text();

  if (text.length === 0) {
    return undefined;
  }

  try {
    return JSON.parse(text) as unknown;
  } catch {
    return text;
  }
}

function extractErrorCode(body: unknown): string | undefined {
  if (!body || typeof body !== 'object') {
    return undefined;
  }

  if ('error' in body && body.error && typeof body.error === 'object') {
    const nested = body.error;

    if ('code' in nested && typeof nested.code === 'string') {
      return nested.code;
    }
  }

  if ('code' in body && typeof body.code === 'string') {
    return body.code;
  }

  if ('error' in body && typeof body.error === 'string') {
    return body.error;
  }

  return undefined;
}

function extractErrorMessage(body: unknown): string | undefined {
  if (!body || typeof body !== 'object') {
    return undefined;
  }

  if ('error' in body && body.error && typeof body.error === 'object') {
    const nested = body.error;

    if ('message' in nested && typeof nested.message === 'string') {
      return nested.message;
    }
  }

  if ('message' in body && typeof body.message === 'string') {
    return body.message;
  }

  if (
    'error_description' in body &&
    typeof body.error_description === 'string'
  ) {
    return body.error_description;
  }

  return undefined;
}

function extractBodyRequestId(body: unknown): string | undefined {
  if (!body || typeof body !== 'object') {
    return undefined;
  }

  if ('error' in body && body.error && typeof body.error === 'object') {
    const nested = body.error;

    if ('requestId' in nested && typeof nested.requestId === 'string') {
      return nested.requestId;
    }
  }

  return undefined;
}

File: packages/python/src/tossinvest_openapi/http.py
from __future__ import annotations

from collections.abc import Mapping
from typing import Any, NamedTuple, Protocol
from urllib.parse import urlencode, urljoin

import requests

from .errors import (
    TossInvestConnectionError,
    TossInvestTimeoutError,
    api_error_from_status,
)
from .types import TossInvestResponseMeta


class HttpRequest(NamedTuple):
    method: str
    url: str
    headers: Mapping[str, str]
    timeout: float
    json: Any = None
    data: Mapping[str, str] | None = None


class HttpSuccess(NamedTuple):
    data: object
    response: TossInvestResponseMeta


class SessionLike(Protocol):
    def request(self, **kwargs: Any) -> Any: ...


def build_url(
    base_url: str,
    path: str,
    query: Mapping[str, object | None] | None = None,
) -> str:
    base = base_url if base_url.endswith("/") else f"{base_url}/"
    url = urljoin(base, path.lstrip("/"))
    defined_query = {key: str(value) for key, value in (query or {}).items() if value is not None}

    if not defined_query:
        return url

    return f"{url}?{urlencode(defined_query)}"


def request_json(session: SessionLike, request: HttpRequest) -> HttpSuccess:
    try:
        response = session.request(
            method=request.method,
            url=request.url,
            headers=dict(request.headers),
            timeout=request.timeout,
            json=request.json,
            data=request.data,
        )
    except requests.Timeout as error:
        raise TossInvestTimeoutError(
            "Toss Invest API request timed out",
            cause=error,
        ) from error
    except requests.RequestException as error:
        raise TossInvestConnectionError(
            "Toss Invest API request failed before receiving a response",
            cause=error,
        ) from error

    body = _parse_json_body(response)
    headers = dict(response.headers)
    request_id = headers.get("x-request-id")
    meta = TossInvestResponseMeta(
        status_code=response.status_code,
        headers=headers,
        request_id=request_id,
    )

    if not response.ok:
        raise api_error_from_status(
            response.status_code,
            message=_extract_error_message(body),
            code=_extract_error_code(body),
            request_id=request_id or _extract_body_request_id(body),
            body=body,
            headers=headers,
        )

    return HttpSuccess(data=body, response=meta)


def _parse_json_body(response: Any) -> object:
    if not response.text:
        return None

    try:
        return response.json()
    except ValueError:
        return response.text


def _extract_error_code(body: object) -> str | None:
    if not isinstance(body, dict):
        return None

    error = body.get("error")
    if isinstance(error, dict):
        code = error.get("code")
        if isinstance(code, str):
            return code
    code = body.get("code")
    if isinstance(code, str):
        return code
    if isinstance(error, str):
        return error
    return None


def _extract_error_message(body: object) -> str | None:
    if not isinstance(body, dict):
        return None

    error = body.get("error")
    if isinstance(error, dict):
        message = error.get("message")
        if isinstance(message, str):
            return message
    message = body.get("message")
    if isinstance(message, str):
        return message
    error_description = body.get("error_description")
    if isinstance(error_description, str):
        return error_description
    return None


def _extract_body_request_id(body: object) -> str | None:
    if not isinstance(body, dict):
        return None

    error = body.get("error")
    if isinstance(error, dict):
        request_id = error.get("requestId")
        if isinstance(request_id, str):
            return request_id
    return None
